File: laundry-reservation/python/tests/in_memory_reservation_repository.py
from __future__ import annotations

from laundry_reservation.reservation import Reservation


class InMemoryReservationRepository:
    def __init__(self) -> None:
        self.all: list[Reservation] = []

    def save(self, reservation: Reservation) -> None:
        for i, r in enumerate(self.all):
            if r.id == reservation.id:
                self.all[i] = reservation
                return
        self.all.append(reservation)

    def find_active_by_customer_email(self, email: str) -> Reservation | None:
        return next(
            (r for r in self.all if r.customer.email == email and not r.is_used),
            None,
        )

    def find_active_by_machine_number(self, machine_number: int) -> Reservation | None:
        return next(
            (r for r in self.all if r.machine_number == machine_number and not r.is_used),
            None,
        )

File: laundry-reservation/typescript/tests/InMemoryReservationRepository.ts
import type { Reservation } from '../src/Reservation.js';
import type { ReservationRepository } from '../src/ReservationRepository.js';

export class InMemoryReservationRepository implements ReservationRepository {
  readonly all: Reservation[] = [];

  save(reservation: Reservation): void {
    const idx = this.all.findIndex(r => r.id === reservation.id);
    if (idx >= 0) {
      this.all[idx] = reservation;
    } else {
      this.all.push(reservation);
    }
  }

  findActiveByCustomerEmail(email: string): Reservation | undefined {
    return this.all.find(r => r.customer.email === email && !r.isUsed);
  }

  findActiveByMachineNumber(machineNumber: number): Reservation | undefined {
    return this.all.find(r => r.machineNumber === machineNumber && !r.isUsed);
  }
}
